pkg/templates/toolchaincluster/template.go

```diff
@@ -0,0 +1,60 @@
+package toolchaincluster
+
+import (
+	"embed"
+
+	v1 "k8s.io/api/core/v1"
+	rbac "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+// GetServiceAccountFromTemplate returns the toolchaincluster service account object from the template content
+func GetServiceAccountFromTemplate(toolchainclusterFS *embed.FS, templateName string, into runtime.Object) (*v1.ServiceAccount, error) {
+	obj, gkv, err := decodeTemplate(toolchainclusterFS, templateName, into)
+	if err != nil {
+		return nil, err
+	}
+	if gkv.Kind == "ServiceAccount" {
+		return obj.(*v1.ServiceAccount), nil
+	}
+	return nil, nil
+}
+
+// GetRoleFromTemplate returns the toolchaincluster Role object from the template content
+func GetRoleFromTemplate(toolchainclusterFS *embed.FS, templateName string, into runtime.Object) (*rbac.Role, error) {
+	obj, gkv, err := decodeTemplate(toolchainclusterFS, templateName, into)
+	if err != nil {
+		return nil, err
+	}
+	if gkv.Kind == "Role" {
+		return obj.(*rbac.Role), nil
+	}
+	return nil, nil
+}
+
+// GetRoleBindingFromTemplate returns the toolchaincluster RoleBinding object from the template content
+func GetRoleBindingFromTemplate(toolchainclusterFS *embed.FS, templateName string, into runtime.Object) (*rbac.RoleBinding, error) {
+	obj, gkv, err := decodeTemplate(toolchainclusterFS, templateName, into)
+	if err != nil {
+		return nil, err
+	}
+	if gkv.Kind == "RoleBinding" {
+		return obj.(*rbac.RoleBinding), nil
+	}
+	return nil, nil
+}
+
+func decodeTemplate(toolchainclusterFS *embed.FS, templateName string, into runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
+	templateContent, err := toolchainclusterFS.ReadFile("templates/toolchaincluster/" + templateName)
+	if err != nil {
+		return nil, nil, err
+	}
+	decode := scheme.Codecs.UniversalDeserializer().Decode
+	obj, gkv, err := decode(templateContent, nil, into)
```

Comment:
with this we technically rely on the fact that there is one manifest per file and that the file name matches the expected kind. Have you considered using kustomize and building the set of objects from it?

Reply:
Do you have an example of using kustomize with embedded files. From what I know we use kustomize to generate static manifests at build time, not runtime. But I'm missing something for sure.

Reply:
Or you are suggesting to keep all objects in one file and generate the individual files at build time ?

Reply:
yeah, you are right with the kustomize, I did not realize it, not sure if they support bundled files, but never mind.
And yeah, the possibility of putting more objects in one file would be nice.
I didn't test it but something like that should work
```go
decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewReader(templateContent), 100)
var objects []*unstructured.Unstructured
for {
	var rawExt runtime.RawExtension
	if err := decoder.Decode(&rawExt); err != nil {
		if err == io.EOF {
			break
		}
		return nil, nil, err
	}
	rawExt.Raw = bytes.TrimSpace(rawExt.Raw)
	if len(rawExt.Raw) == 0 || bytes.Equal(rawExt.Raw, []byte("null")) {
		continue
	}
	unstructuredObj := &unstructured.Unstructured{}
	_, _, err = scheme.Codecs.UniversalDeserializer().Decode(rawExt.Raw, nil, unstructuredObj)
	if err != nil {
		return nil, nil, err
	}
	objects = append(objects, unstructuredObj)
}
```

Reply:
and btw, with this approach we don't need to rely on the filenames matching the kinds anymore - the logic can just read all the files from the folder and collect all objects that are defined in the files.

Reply:
Nice, I was not aware of this usage of `runtime.RawExtension`. Thanks!

I've implemented this and moved all the files in the same template. I've updated also the logic to scan recursively for all files in the given directory.